src/theory/quantifiers/sygus/sygus_reconstruct.h

```diff
@@ -0,0 +1,300 @@
+/*********************                                                        */
+/*! \file sygus_reconstruct.h
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Abdalrhman Mohamed
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2021 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief utility for reconstructing terms to match a grammar
+ **/
+
+#include "cvc4_private.h"
+
+#ifndef CVC4__THEORY__QUANTIFIERS__SYGUS_RECONSTRUCT_H
+#define CVC4__THEORY__QUANTIFIERS__SYGUS_RECONSTRUCT_H
+
+#include <map>
+#include <vector>
+
+#include "theory/quantifiers/sygus/rcons_obligation_info.h"
+#include "theory/quantifiers/sygus/rcons_type_info.h"
+
+namespace CVC4 {
+namespace theory {
+namespace quantifiers {
+
+using ObligationSet = std::unordered_set<Node, NodeHashFunction>;
+using TypeObligationSetMap =
+    std::unordered_map<TypeNode, ObligationSet, TypeNodeHashFunction>;
+
+/** SygusReconstruct
+ *
+ * This class implements an algorithm for reconstructing a given term such that
+ * the reconstructed term is equivalent to the original term and its syntax
+ * matches a specified grammar.
+ *
+ * Goal: find a term g in sygus type T that is equivalent to builtin term t.
+ *
+ * rcons(t_0, T_0) returns g
+ * {
+ *   Obs: A map from sygus types T to a set of triples to reconstruct into T,
+ *        where each triple is of the form (k, t, s), where k is a skolem of
+ *        type T, t is a builtin term of the type encoded by T, and s is a
+ *        possibly null sygus term of type T representing the solution.
+ *
+ *   Sol: A map from a skolems k to slutions of their corresponding obligations
+ *        such that Obs[T] contains entries (k, t, Sol[k]).
```

Comment:
Say what is the obligation of `k`. Define `Obs` and `Sol`

Reply:
By define `Obs` and `Sol` , do you mean give them an initial value?

Reply:
No, I meant to say what they are. But you already had and I missed it, sorry.